allauth/socialaccount/providers/oauth2/tests.py

```diff
@@ -0,0 +1,147 @@
+                                       name=FakeOAuth2Adapter.provider_id,
+                                       client_id='app123id',
+                                       key=FakeOAuth2Adapter.provider_id,
+                                       secret='dummy')
+        app.sites.add(get_current_site())
+
+
+class OAuth2TestsNoProxying(OAuth2Tests):
+    def setUp(self):
+        self.init_request('fake_login', dict(process='login'))
+        super(OAuth2TestsNoProxying, self).setUp()
+
+    def test_proxyless_login(self):
+        login_view = OAuth2LoginView.adapter_view(FakeOAuth2Adapter)
+        login_response = login_view(self.request)
+        self.assertEqual(login_response.status_code, 302)  # Redirect
+        self.assertEqual(self.param('redirect_uri', login_response['location']),
+                         'http://testserver/fake/login/callback/')
+
+    def test_is_not_login_proxy(self):
+        with self.assertRaises(NoReverseMatch):
+            reverse('fake_proxy')
+
+
+@override_settings(ACCOUNT_LOGIN_CALLBACK_PROXY='https://loginproxy')
+class OAuth2TestsUsesProxy(OAuth2Tests):
+    def setUp(self):
+        self.init_request('fake_login', dict(process='login'))
+        super(OAuth2TestsUsesProxy, self).setUp()
+
+    def test_login_by_proxy(self):
+        login_view = OAuth2LoginView.adapter_view(FakeOAuth2Adapter)
+        login_response = login_view(self.request)
+        self.assertEqual(login_response.status_code, 302)  # Redirect
+        self.assertEqual(self.param('redirect_uri', login_response['location']),
+                         'https://loginproxy/fake/login/callback/proxy/')
+        state = json.loads(self.param('state', login_response['location']))
+        self.assertEqual(state['host'], 'http://testserver/fake/login/')
+
+    def test_is_not_login_proxy(self):
+        with self.assertRaises(NoReverseMatch):
+            reverse('fake_proxy')
+
+@override_settings(ACCOUNT_LOGIN_PROXY_REDIRECT_WHITELIST=
+                       'https://cheshirecat,https://tweedledee,')
+class OAuth2TestsIsProxy(OAuth2Tests):
+    def reload_urls(self):
+        for module in sys.modules:
+            if module.endswith('urls'):
+                reload(sys.modules[module])
+        clear_url_caches()
+
+
+    def setUp(self):
+        super(OAuth2TestsIsProxy, self).setUp()
+        self.reload_urls()
+
+    @override_settings(ACCOUNT_LOGIN_PROXY_REDIRECT_WHITELIST='')
+    def tearDown(self):
+        super(OAuth2TestsIsProxy, self).tearDown()
+        self.reload_urls()
+
+    def tests_is_login_proxy(self):
+        reverse('fake_proxy')
+
+    def test_rejects_request_with_no_host_in_state(self):
+        self.init_request('fake_proxy', dict(process='login'))
+        with self.assertRaises(MissingParameter):
+            proxy_login_callback(
+                self.request, callback_view_name='fake_callback')
+
+    def test_rejects_request_with_unwhitelisted_host(self):
```

Comment:
(i don't know that that case ultimately matters much, but i'd just like to know what it does in that case)